src/feature/hs/hs_client.c

```diff
@@ -576,10 +576,21 @@ send_introduce1(origin_circuit_t *intro_circ,
   /* Send the INTRODUCE1 cell. */
   if (hs_circ_send_introduce1(intro_circ, rend_circ, ip,
                               desc->subcredential) < 0) {
-    /* Unable to send the cell, the intro circuit has been marked for close so
-     * this is a permanent error. */
-    tor_assert_nonfatal(TO_CIRCUIT(intro_circ)->marked_for_close);
-    goto perm_err;
+    if (TO_CIRCUIT(intro_circ)->marked_for_close) {
```

Comment:
Why not do this in `hs_circ_send_introduce1()` instead of clobbering this function with this info, and trying to infer what `hs_circ_send_introduce1()` has done?

Reply:
Good question. The reason is that I wanted that to be a "client action" and not part of the "HS circuit" layer. The client there decides to note down the intro point client cache failure, not the circuit layer.

In other word, I wanted to not have `hs_circuit` to callback into `hs_client`.